src/geometry/euclidean.lean

```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+    { rw norm_eq_zero at hx,
+      rw [hx, inner_zero_left] },
+    { rw norm_eq_zero at hy,
+      rw [hy, inner_zero_right] } },
+  { exact div_mul_cancel _ h }
+end
+
+/-- The sine of the angle between two vectors, multiplied by the
+product of their norms. -/
+lemma sin_angle_of_vectors_mul_norm_mul_norm (x y : V) :
+  (angle_of_vectors x y).sin * (∥x∥ * ∥y∥) =
+    real.sqrt (inner x x * inner y y - inner x y * inner x y) :=
+begin
+  unfold angle_of_vectors,
+  rw [real.sin_arccos (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).1
+                     (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).2,
+      ←real.sqrt_mul_self (mul_nonneg (norm_nonneg x) (norm_nonneg y)),
+      ←real.sqrt_mul' _ (mul_self_nonneg _), pow_two,
+      real.sqrt_mul_self (mul_nonneg (norm_nonneg x) (norm_nonneg y)), inner_self_eq_norm_square,
+      inner_self_eq_norm_square],
+  by_cases h : (∥x∥ * ∥y∥) = 0,
+  { rw [(show ∥x∥ * ∥x∥ * (∥y∥ * ∥y∥) = (∥x∥ * ∥y∥) * (∥x∥ * ∥y∥), by ring), h, mul_zero, mul_zero,
+        zero_sub],
+    cases eq_zero_or_eq_zero_of_mul_eq_zero h with hx hy,
+    { rw norm_eq_zero at hx,
+      rw [hx, inner_zero_left, zero_mul, neg_zero] },
+    { rw norm_eq_zero at hy,
+      rw [hy, inner_zero_right, zero_mul, neg_zero] } },
+  { field_simp [h],
+    ring }
+end
+
+/-- The angle between two vectors is zero if and only if they are
+nonzero and one is a positive multiple of the other. -/
+lemma angle_of_vectors_eq_zero_iff (x y : V) :
+  angle_of_vectors x y = 0 ↔ (x ≠ 0 ∧ ∃ (r : ℝ), 0 < r ∧ y = r • x) :=
+begin
+  unfold angle_of_vectors,
+  rw [←inner_div_norm_mul_norm_eq_one_iff, ←real.arccos_one],
+  split,
+  { intro h,
+    exact real.arccos_inj (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).1
+                          (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).2
+                          (by norm_num)
+                          (by norm_num)
+                          h },
+  { intro h,
+    rw h }
+end
+
+/-- The angle between two vectors is π if and only if they are nonzero
+and one is a negative multiple of the other. -/
+lemma angle_of_vectors_eq_pi_iff (x y : V) :
+  angle_of_vectors x y = real.pi ↔ (x ≠ 0 ∧ ∃ (r : ℝ), r < 0 ∧ y = r • x) :=
+begin
+  unfold angle_of_vectors,
+  rw [←inner_div_norm_mul_norm_eq_neg_one_iff, ←real.arccos_neg_one],
+  split,
+  { intro h,
+    exact real.arccos_inj (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).1
+                          (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).2
+                          (by norm_num)
+                          (by norm_num)
+                          h },
+  { intro h,
+    rw h }
+end
+
+/-- If the angle between two vectors is π, the angles between those
+vectors and a third vector add to π. -/
+lemma angle_of_vectors_add_angle_of_vectors_eq_pi_of_angle_of_vectors_eq_pi {x y : V} (z : V)
+    (h : angle_of_vectors x y = real.pi) :
+  angle_of_vectors x z + angle_of_vectors y z = real.pi :=
```

Comment:
I know it doesn't matter for the maths, but would `angle z y` for the second summand be slightly easier in applications? You probably know this best...

Reply:
I gave this one in symmetric form, but I suspect what's easier for applications is usually directed angles not these undirected angles at all (directed angles are more complicated to define, first you need to define orientations of finite-dimensional vector spaces / subspaces over ordered fields = equivalence classes of bases where two bases are equivalent if the map between them has positive determinant).